File: collect_app/src/main/java/org/odk/collect/android/formmanagement/FormDownloadExceptionMapper.java
package org.odk.collect.android.formmanagement;

import android.content.Context;

import org.odk.collect.android.R;
import org.odk.collect.android.utilities.TranslationHandler;

public class FormDownloadExceptionMapper {

    private final Context context;

    public FormDownloadExceptionMapper(Context context) {
        this.context = context;
    }

    public String getMessage(FormDownloadException exception) {
        switch (exception.getType()) {
            case DUPLICATE_FORMID_VERSION:
                // Not localized as we figure out how common this is and whether the wording is helpful
                return "You've already downloaded a form with the same ID and version but with different contents. " +
                        "Before downloading, please send all data you have collected with the existing form and delete the data and blank form.";
            case GENERIC:
            default:
                return TranslationHandler.getString(context, R.string.failure);
        }
    }
}

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/FormDownloadExceptionMapper.kt
package org.odk.collect.android.formmanagement

import android.content.Context
import org.odk.collect.android.R
import org.odk.collect.strings.localization.getLocalizedString

class FormDownloadExceptionMapper(private val context: Context) {
    fun getMessage(exception: FormDownloadException?): String {
        return when (exception) {
            is FormDownloadException.FormWithNoHash -> {
                context.getLocalizedString(
                    org.odk.collect.strings.R.string.form_with_no_hash_error
                ) + " " + context.getLocalizedString(
                    org.odk.collect.strings.R.string.report_to_project_lead
                )
            }
            is FormDownloadException.FormParsingError -> {
                context.getLocalizedString(
                    org.odk.collect.strings.R.string.form_parsing_error
                ) + " " + context.getLocalizedString(
                    org.odk.collect.strings.R.string.report_to_project_lead
                )
            }
            is FormDownloadException.DiskError -> {
                context.getLocalizedString(
                    org.odk.collect.strings.R.string.form_save_disk_error
                ) + " " + context.getLocalizedString(
                    org.odk.collect.strings.R.string.report_to_project_lead
                )
            }
            is FormDownloadException.InvalidSubmission -> {
                context.getLocalizedString(
                    org.odk.collect.strings.R.string.form_with_invalid_submission_error
                ) + " " + context.getLocalizedString(
                    org.odk.collect.strings.R.string.report_to_project_lead
                )
            }
            is FormDownloadException.FormSourceError -> {
                FormSourceExceptionMapper(context).getMessage(exception.exception)
            }
            else -> {
                context.getLocalizedString(org.odk.collect.strings.R.string.report_to_project_lead)
            }
        }
    }
}
